File: sdk/go/compact_test.go
package microsandbox

import (
	"context"
	"encoding/json"
	"testing"
)

func TestCompactionPreservesExplicitZeroAndRejectsCloudHandle(t *testing.T) {
	zero := uint32(0)
	raw, err := json.Marshal(DiskCompactionOptions{Layers: &zero})
	if err != nil || string(raw) != `{"layers":0}` {
		t.Fatalf("explicit zero lost: %s %v", raw, err)
	}
	h := &SandboxHandle{name: "same-name-as-local", backendKind: BackendCloud}
	if _, err := h.Compact(context.Background(), DiskCompactionOptions{}); !IsKind(err, ErrUnsupportedOperation) {
		t.Fatalf("cloud handle must refuse before local lookup: %v", err)
	}
}

func TestCompactionSelectorsAndDiskResults(t *testing.T) {
	encoded, err := json.Marshal(DiskCompactionOptions{Disk: "/data", RootDiskOnly: true})
	if err != nil || string(encoded) != `{"disk":"/data","root_disk_only":true}` {
		t.Fatalf("selectors must reach shared validation without being dropped: %s %v", encoded, err)
	}
	result, err := parseCompaction(`{"dry_run":true,"input_layers":6,"selected_layers":4,"output_layers":4,"materialized_bytes":0,"total_us":10,"pause_us":0,"disks":[{"guest_path":"/data","input_layers":3,"selected_layers":2,"output_layers":2,"materialized_bytes":0,"total_us":4}]}`, nil)
	if err != nil || len(result.Disks) != 1 || result.Disks[0].GuestPath != "/data" || result.Disks[0].SelectedLayers != 2 {
		t.Fatalf("per-disk result lost: %#v %v", result, err)
	}
}

File: sdk/node-ts/tests/unit/compact.test.ts
import { describe, expect, it, vi } from "vitest";

// CI runs these tests from build artifacts without the TypeScript source tree.
import { compactionResultFromJson, type DiskCompactionOptions } from "../../dist/compact.js";
import { Sandbox } from "../../dist/sandbox.js";
import { SandboxHandle } from "../../dist/sandbox-handle.js";
import type { NapiSandbox, NapiSandboxHandle } from "../../dist/internal/napi.js";

describe("disk compaction contract", () => {
  it("retains per-disk outcomes alongside aggregate metrics", () => {
    const result = compactionResultFromJson(JSON.stringify({
      dry_run: true, input_layers: 6, selected_layers: 4, output_layers: 4,
      materialized_bytes: 0, total_us: 10, pause_us: 0,
      disks: [{ guest_path: "/data", input_layers: 3, selected_layers: 2,
        output_layers: 2, materialized_bytes: 0, total_us: 4 }],
    }));
    expect(result.disks).toEqual([{ guestPath: "/data", inputLayers: 3,
      selectedLayers: 2, outputLayers: 2, materializedBytes: 0, totalUs: 4 }]);
    const options: DiskCompactionOptions = { disk: "/data", rootDiskOnly: false, layers: 999 };
    expect(options.disk).toBe("/data");
  });

  it("represents an empty default selection", () => {
    expect(compactionResultFromJson('{"disks":[]}').disks).toEqual([]);
  });

  it("forwards both selectors for shared validation without dropping either", async () => {
    const compact = vi.fn(async () => '{"disks":[]}');
    const sandbox = new Sandbox({ compact } as unknown as NapiSandbox, "worker");
    await sandbox.compact({ layers: 99, disk: "/data", rootDiskOnly: true, dryRun: true });
    expect(compact).toHaveBeenCalledWith(99, true, "/data", true);

    const handle = new SandboxHandle({ compact } as unknown as NapiSandboxHandle);
    await handle.compact({ disk: "/" });
    expect(compact).toHaveBeenLastCalledWith(undefined, undefined, "/", undefined);
    await handle.compact();
    expect(compact).toHaveBeenLastCalledWith(undefined, undefined, undefined, undefined);
  });
});
